plugin/gcp/firestoreDocument.go

```diff
@@ -0,0 +1,118 @@
+package gcp
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"cloud.google.com/go/firestore"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+type firestoreDocument struct {
+	plugin.EntryBase
+	client firestoreProjectClient
+	path   string
+	data   map[string]interface{}
+}
+
+type firestoreDocumentMetadata struct {
+	CreateTime time.Time              `json:"CreateTime"`
+	UpdateTime time.Time              `json:"UpdateTime"`
+	ReadTime   time.Time              `json:"ReadTime"`
+	Data       map[string]interface{} `json:"Data"`
+}
+
+func newFirestoreDocument(client firestoreProjectClient, parent string, snapshot *firestore.DocumentSnapshot) *firestoreDocument {
+	doc := &firestoreDocument{
+		EntryBase: plugin.NewEntry(snapshot.Ref.ID),
+		client:    client,
+		path:      firestorePath(parent, snapshot.Ref.ID),
+		data:      snapshot.Data(),
+	}
+
+	doc.Attributes().
+		SetCrtime(snapshot.CreateTime).
+		SetCtime(snapshot.UpdateTime).
+		SetMtime(snapshot.UpdateTime).
+		SetMeta(firestoreDocumentMetadata{
+			snapshot.CreateTime,
+			snapshot.UpdateTime,
+			snapshot.ReadTime,
+			snapshot.Data(),
+		})
+
+	return doc
+}
+
+func (doc *firestoreDocument) List(ctx context.Context) ([]plugin.Entry, error) {
+	colls, err := doc.client.Doc(doc.path).Collections(ctx).GetAll()
+	if err != nil {
+		return nil, err
+	}
+	dataJSON, err := newFirestoreDocumentDataJSON(doc.data)
+	if err != nil {
+		return nil, err
+	}
+	collEntries := toCollectionEntries(doc.client, doc.path, colls)
+	return append([]plugin.Entry{dataJSON}, collEntries...), nil
+}
+
+func (doc *firestoreDocument) Delete(ctx context.Context) (bool, error) {
+	_, err := doc.client.Doc(doc.path).Delete(ctx)
```

Comment:
Will this result in deleting all the collections under this document?

Reply:
Yup.